packages/cardano-services-client/src/WebSocket.ts

```diff
@@ -0,0 +1,182 @@
+
+export type WSHandler = (message: WSMessage) => void;
+
+export interface WsClientConfiguration {
+  /** The interval in seconds between two heartbeat messages. Default 55". */
+  heartbeatInterval?: number;
+
+  /** The interval in seconds after which a request must timeout. Default 60". */
+  requestTimeout?: number;
+
+  /** The WebSocket server URL. */
+  url: URL;
+}
+
+export interface WsClientDependencies {
+  /** The logger. */
+  logger: Logger;
+}
+
+const isEventError = (error: unknown): error is { error: Error } =>
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  typeof error === 'object' && !!error && (error as any).error instanceof Error;
+
+export class CardanoWsClient {
+  /** The client id, assigned by the server. */
+  clientId = NOT_CONNECTED_ID;
+
+  /** The `Observable` form of `NetworkInfoProvider`. */
+  networkInfo: {
+    [m in NetworkInfoMethods]: Observable<AsyncReturnType<NetworkInfoProvider[m]>>;
+  };
+
+  /** WebSocket based `NetworkInfoProvider` implementation. */
+  networkInfoProvider: NetworkInfoProvider;
+
+  private heartbeatInterval: number;
+  private heartbeatTimeout: NodeJS.Timeout;
+  private logger: Logger;
+  private status: WSStatus = 'idle';
+  private url: URL;
+  private ws: WebSocket;
+
+  private networkInfoSubjects = {} as {
+    [m in NetworkInfoMethods]: ReplaySubject<AsyncReturnType<NetworkInfoProvider[m]>>;
+  };
+
+  constructor(deps: WsClientDependencies, cfg: WsClientConfiguration) {
+    this.heartbeatInterval = (cfg.heartbeatInterval || 55) * 1000;
+    this.logger = deps.logger;
+    this.url = cfg.url;
+
+    this.networkInfoSubjects = {
+      eraSummaries: new ReplaySubject<EraSummary[]>(1),
+      genesisParameters: new ReplaySubject<Cardano.CompactGenesis>(1),
+      ledgerTip: new ReplaySubject<Cardano.Tip>(1),
+      lovelaceSupply: new ReplaySubject<SupplySummary>(1),
+      protocolParameters: new ReplaySubject<Cardano.ProtocolParameters>(1),
+      stake: new ReplaySubject<StakeSummary>(1)
+    };
+    this.networkInfo = this.networkInfoSubjects;
+
+    this.networkInfoProvider = {
+      eraSummaries: () => firstValueFrom(this.networkInfo.eraSummaries),
+      genesisParameters: () => firstValueFrom(this.networkInfo.genesisParameters),
+      healthCheck: () => Promise.resolve({ ok: this.status === 'connected' }),
+      ledgerTip: () => firstValueFrom(this.networkInfo.ledgerTip),
+      lovelaceSupply: () => firstValueFrom(this.networkInfo.lovelaceSupply),
+      protocolParameters: () => firstValueFrom(this.networkInfo.protocolParameters),
+      stake: () => firstValueFrom(this.networkInfo.stake)
+    };
+
+    this.connect();
+  }
+
+  private connect() {
+    this.status = 'connecting';
+
+    const ws = new WebSocket(this.url);
+
+    // eslint-disable-next-line sonarjs/cognitive-complexity
+    ws.onmessage = (event) => {
+      try {
+        if (typeof event.data !== 'string') throw new Error('Unexpected data from WebSocket ');
```

Comment:
This should never happen. I think for now nothing more than a log is required.

In case it will actually never happen, nothing more is required, in case it happens, we should check what causes it in how to handle.

Reply:
Should we `logger.error` instead of throw then?

Reply:
It is, in the `catch` block the error is logged; it is also thrown as being not a string the message can't be parsed so next steps can't be performed